Define core networking for User Services

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##
This PR implements the core networking for the User Services account, specifically the VPC, subnets, NAT/Internet gateways, routing, and Transit Gateway attachment.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
This repository is meant to contain all networking components for User Services, similar to what we do in [cisagov/cool-sharedservices-networking](https://github.com/cisagov/cool-sharedservices-networking).

This work is part of https://github.com/cisagov/cool-system/issues/114.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
All changes were successfully applied in the Staging environment.

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All new and existing tests pass.

dav3r committed:
Remove unused "examples" directory inherited from skeleton

dav3r committed:
Trim list of code owners

dav3r committed:
Remove unused file inherited from skeleton

dav3r committed:
Configure Terraform backend

dav3r committed:
Define initial set of variables

dav3r committed:
Set up providers and local variables

dav3r committed:
Add all needed remote TF states

dav3r committed:
Define policy that can provision networking and attach it to ProvisionAccount role in User Services

dav3r committed:
Create VPC and internet gateway

dav3r committed:
Enable VPC flow logs

dav3r committed:
Create subnets and NAT gateways

dav3r committed:
Attach User Services VPC to the Shared Services TGW

dav3r committed:
Create route tables and routes

dav3r committed:
Define initial set of outputs

dav3r committed:
Update README from skeleton

jsf9k commented on outputs.tf:
```diff
@@ -1,24 +1,24 @@
-output "private_ip" {
-  value       = aws_instance.example.private_ip
-  description = "The private IP of the EC2 instance"
+output "transit_gateway_id" {
+  value       = local.transit_gateway_id
+  description = "The ID of the Transit Gateway in the Shared Services account."
```
This probably doesn't need to be an output, since it's already an output of [cisagov/cool-sharedservices-networking](https://github.com/cisagov/cool-sharedservices-networking).

dav3r replied:
Good call- I removed the output in https://github.com/cisagov/cool-userservices-networking/pull/1/commits/ffb1e583b0b03bc42adfa499031533d4d6ff937c.

jsf9k reviewed (changes requested):
I think this looks good, except for one comment.  Strong work!

dav3r committed:
Remove output for Transit Gateway

As pointed out by @jsf9k, this output is already available from 
cool-sharedservices-networking.

dav3r committed:
Remove unnecessary local variable

It is not needed since User Services utilizes the TGW's default route 
table.

jsf9k reviewed (approved):
Great work!

jsf9k commented on transitgateway.tf:
```diff
@@ -0,0 +1,26 @@
+  transit_gateway_id = local.transit_gateway_id
+  vpc_id             = aws_vpc.userservices.id
+}
+
+# Add a route to the User Services VPC in the TGW default route table
+resource "aws_ec2_transit_gateway_route" "userservices_route" {
+  provider = aws.sharedservicesprovisionaccount
+
+  destination_cidr_block         = aws_vpc.userservices.cidr_block
+  transit_gateway_attachment_id  = aws_ec2_transit_gateway_vpc_attachment.userservices.id
+  transit_gateway_route_table_id = local.transit_gateway_default_route_table_id
+}
```
This block is unnecessary and should be removed.  The route will be automatically generated and propagated to the default route table, which is preferable to static routing.
```suggestion
```

You should, however, add a route to User Services to the _other (non-default)_ route tables, similar to what we do [here for Shared Services](https://github.com/cisagov/cool-sharedservices-networking/blob/develop/transitgateway.tf#L74-L84).

The idea is that Shared Services and User Services can get to any account via the TGW, but other environments are restricted and can only access the Shared Services and User Services accounts.

dav3r replied:
Excellent catches!  Good thing you noticed that now- it would've taken me a while to troubleshoot this later when we first attempt sending traffic to the User Services VPC.

Please take a look at https://github.com/cisagov/cool-userservices-networking/pull/1/commits/d40c009445e7b12c53751a51d2c03c5f7cd7bcb3 and https://github.com/cisagov/cool-userservices-networking/pull/1/commits/6aec5ef6e2fb5d33a6a648f443a965d9709feb72 to ensure that I didn't misunderstand you.

jsf9k reviewed (changes requested):
Please see my comment in `transitgateway.tf`.

dav3r committed:
Remove unnecessary route in TGW default route table

As pointed out by @jsf9k, this route will be automatically generated and 
propagated to the default route table, which is preferable to static 
routing.

dav3r committed:
Add route to User Services VPC to the route tables of other accounts that may require it

At the moment, this is the env* and the PCA accounts.

jsf9k reviewed (approved):
Thanks for making that change.  Everything looks correct now.

hillaryj reviewed (approved):
🎉

mcdonnnj reviewed (approved):
Progress! 💪
